Finalizes antagonist datums, migrates existing gamemodes

Xhuis opened the pull request:
___This is WIP, please don't merge it!___

The goal of this PR is to finalize antagonist datums by both improving their function and by moving most existing gamemode code to be handled by them. Gamemodes will still exist for probability, assignment, and things like that.

An overhaul of gamemodes might come later, but it isn't in the scope of this PR. Anything in the code marked with (TBWO) has yet to be ported over to the new system, so remind me if I forget about it.

Commit:
Antagonist datums, second iteration

ChangelingRain commented on code/datums/antagonists/antag_datum.dm:
```diff
@@ -1,35 +1,64 @@
+/datum/antagonist/proc/forge_objectives()
+	if(!objective_types.len)
+		return
+	for(var/i = 1, i > number_of_objectives, i++)
```
for(var/i in 1 to number_of_objectives)

Xhuis replied:
Sorry, I'm still rusty!

ChangelingRain commented on code/datums/antagonists/antag_datum.dm:
```diff
@@ -73,6 +102,12 @@
 	var/datum/antagonist/D = new datum_type()
 	. = D.give_to_body(src)
 
+/mob/living/proc/lose_antag_datum(datum_type) //tries to remove an antagonist datum from a mob
```
should have the same 'check_subtypes' arg as below, so that if something uses subtypes for specific effects you don't need to try and lose each type, just the main type.

ChangelingRain commented on code/datums/antagonists/antag_datum.dm:
```diff
@@ -1,35 +1,64 @@
 /datum/antagonist/proc/can_be_owned(mob/living/new_body)
-	return new_body && !new_body.has_antag_datum(prevented_antag_datum_type, TRUE)
+	for(var/D in prevented_antag_datum_types)
+		if(D == type)
```
there should be a var or something so this can also check subtypes

JJRcop commented:
Isn't @AnturK doing this?

Commit:
First work on objectives

Xhuis commented:
Hopefully I can shift that opinion around. Once I finish with this, gamemode code will be much more optimized, and it will be much, much easier to expand the system for new antagonists.

Commit:
Finalizes the traitor datum

lordpidey commented:
Hey, I was working on #21217, before I got sidetracked.  Would you like me to help you include that in here?

Xhuis commented:
After a few attempted and failed forays into changing traitor code to fit the new system, I've discovered that a combination of bad code and extreme age make gamemode code very heavily rooted in the game's code. Updating and refining it will be a matter of extreme patience, meticulousness, and debugging.

I'm actually going to go as far as to request a gamemode code freeze until this is merged - otherwise, it might take several months before I finish this.

@lordpidey Although I'd appreciate the help, I would have to do a lot of refactoring to get what you've already done to work properly. Still, I'll look into porting the code over (crediting you, of course!)

lordpidey commented:
Okay, thank you @Xhuis.

When I wrote devils, I kept in mind that datum antags would be coming soon, and wrote it so that it could be ported easily.

However, I made one mistake.  I attached devil to a mind rather than a mob, thinking that datum antags would be attached to mobs.  My mistake.

Commit:
Finishes the traitor datum

Xhuis commented:
__In this commit:__
* Temporarily removed double agents and exchange objectives. They will be returned later.
* Added extra infrastructure in antag datums for gain and loss fluff text.
* Added new framework to call procs on antag datums remotely.
* Wrapped up the last of the traitor datum.

Commit:
More finishing, beginning work on changeling datum

ChangelingRain commented on code/datums/antagonists/antag_datum.dm:
```diff
@@ -83,8 +89,9 @@
 	owner.antag_datums -= src
-	owner = new_body
-	apply_innate_effects()
+	spawn(1) //Give the game time to sort out new minds, bodies...
```
this shouldn't be needed, as it should only be called after the mind has been transferred to the new body

KorPhaeron commented:
stale, let me know when you're ready
